collector/ccr_stats.go

```diff
@@ -0,0 +1,448 @@
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.OperationsWritten)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "time_since_last_read_millis"),
+					"The number of milliseconds since a read request was sent to the leader",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.TimeSinceListReadMillis)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "replication_lag"),
+					"Represents value of how much follower lagging the leader. Based on difference between leader_global_checkpoint and follower_global_checkpoint",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.LeaderGlobalCheckpoint - indexCCRStatsShard.FollowerGlobalCheckpoint)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+		},
+	}
+}
+
+func (s *CCRStats) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range s.ccrMetrics {
+		ch <- metric.Desc
+	}
+	ch <- s.up.Desc()
+	ch <- s.totalScrapes.Desc()
+	ch <- s.jsonParseFailures.Desc()
+}
+
+func (s *CCRStats) fetchCCRStatsResponse() (CCRStatsResponse, error) {
+	var ccrResponse CCRStatsResponse
+
+	u := *s.url
+	u.Path = path.Join(u.Path, "/_all/_ccr/stats")
+
+	res, err := s.client.Get(u.String())
+	if err != nil {
+		return ccrResponse, fmt.Errorf("failed to get ccr statistics from %s://%s:%s%s: %s",
+			u.Scheme, u.Hostname(), u.Port(), u.Path, err)
+	}
+
+	defer func() {
+		err = res.Body.Close()
+		if err != nil {
+			_ = level.Warn(s.logger).Log(
+				"msg", "failed to close http.Client",
+				"err", err,
+			)
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		return ccrResponse, fmt.Errorf("HTTP Request failed with code %d", res.StatusCode)
+	}
+
+	bts, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		s.jsonParseFailures.Inc()
+		return ccrResponse, err
+	}
+
+	if err := json.Unmarshal(bts, &ccrResponse); err != nil {
+		s.jsonParseFailures.Inc()
+		return ccrResponse, err
+	}
+	return ccrResponse, err
+}
+
+//Collect gets CCRStats metric values
+func (s *CCRStats) Collect(ch chan<- prometheus.Metric) {
+
+	s.totalScrapes.Inc()
+	defer func() {
+		ch <- s.up
+		ch <- s.totalScrapes
+		ch <- s.jsonParseFailures
+	}()
+
+	indexCCRStatsResponse, err := s.fetchCCRStatsResponse()
+	if err != nil {
+		s.up.Set(0)
+		_ = level.Warn(s.logger).Log(
+			"msg", "failed to fetch and decode CCRStats stats",
+			"err", err,
+		)
+		return
+	}
+	s.up.Set(1)
+
+	for _, indexCCRStatsResponse := range indexCCRStatsResponse.IndexCCRStats {
```

Comment:
When these metrics are package vars, we can call them out explicitly here instead of using a loop.